corehq/apps/api/object_fetch_api.py

```diff
@@ -54,58 +72,109 @@ def get(self, *args, **kwargs):
+                    fixed_size=size
+                )
         else:
-            # default stream
-            attachment_meta, attachment_stream = CommCareCase.fetch_case_attachment(case_id, attachment_key)
+            cached_attachment = get_cached_case_attachment(domain, case_id, attachment_id)
+            attachment_meta, attachment_stream = cached_attachment.get()
 
         if attachment_meta is not None:
             mime_type = attachment_meta['content_type']
         else:
             mime_type = "plain/text"
-        response = StreamingHttpResponse(streaming_content=attachment_stream,
-                content_type=mime_type)
-        return response
+
+        return StreamingHttpResponse(streaming_content=attachment_stream, content_type=mime_type)
 
 
 class FormAttachmentAPI(View):
     @method_decorator(login_or_digest_or_basic_or_apikey())
-    def get(self, *args, **kwargs):
-        form_id = kwargs.get('form_id', None)
-        if not form_id or not XFormInstance.get_db().doc_exist(form_id):
+    def get(self, request, domain, form_id=None, attachment_id=None):
+        if not form_id or not attachment_id or not FormAccessors(domain).form_exists(form_id):
             raise Http404
 
-        attachment_key = kwargs.get('attachment_id', None)
-
         try:
-            resp = XFormInstance.get_db().fetch_attachment(form_id, attachment_key, stream=True)
-        except ResourceNotFound:
+            content = FormAccessors(domain).get_attachment_content(form_id, attachment_id)
+        except AttachmentNotFound as e:
+            print e
             raise Http404
         
-        headers = resp.resp.headers
-        content_type = headers.get('Content-Type', None)
+        return StreamingHttpResponse(streaming_content=content.content_stream, content_type=content.content_type)
+
+
+def fetch_case_image(domain, case_id, attachment_id, filesize_limit=0, width_limit=0, height_limit=0, fixed_size=None):
+    """
+    Return (metadata, stream) information of best matching image attachment.
+
+    :param attachment_id: the identifier of the attachment to fetch
+    """
+    if fixed_size is not None:
+        size_key = fixed_size
+    else:
+        size_key = OBJECT_ORIGINAL
+
+    constraint_dict = {}
+    if filesize_limit:
+        constraint_dict['content_length'] = filesize_limit
+
+    if height_limit:
+        constraint_dict['height'] = height_limit
+
+    if width_limit:
+        constraint_dict['width'] = width_limit
+    do_constrain = bool(constraint_dict)
+
+    cached_image = get_cached_case_attachment(domain, case_id, attachment_id, is_image=True)
+    meta, stream = cached_image.get(size_key=size_key)
+
+    if do_constrain:
+        def meets_constraint(constraints, meta):
+            for c, limit in constraints.items():
+                if meta[c] > limit:
+                    return False
+            return True
+
+        if meets_constraint(constraint_dict, meta):
```

Comment:
this seems a little odd, maybe just have `if not meets_constraint(constraint_dict, meta):`